scylla/src/transport/query_result.rs

```
@@ -0,0 +1,210 @@
+    pub fn rows<'s, R: DeserializeRow<'s>>(&'s self) -> Result<TypedRowIterator<'s, R>, RowsError> {
+        Ok(self
+            .raw_rows
+            .as_ref()
+            .ok_or(RowsError::NotRowsResponse)?
+            .rows_iter()?)
+    }
+
+    /// Returns the received rows when present, or None.
+    ///
+    /// Returns an error if the rows are present but are of incorrect type.
+    pub fn maybe_rows<'s, R: DeserializeRow<'s>>(
+        &'s self,
+    ) -> Result<Option<TypedRowIterator<'s, R>>, ParseError> {
+        match &self.raw_rows {
+            Some(rows) => Ok(Some(rows.rows_iter()?)),
+            None => Ok(None),
+        }
+    }
+
+    /// Returns `Ok` for a result of a query that shouldn't contain any rows.\
+    /// Will return `Ok` for `INSERT` result, but a `SELECT` result, even an empty one, will cause an error.\
+    /// Opposite of [`rows()`](QueryResult::rows).
+    pub fn result_not_rows(&self) -> Result<(), ResultNotRowsError> {
+        match &self.raw_rows {
+            Some(_) => Err(ResultNotRowsError),
+            None => Ok(()),
+        }
+    }
+
+    /// Returns first row from the received rows.\
+    /// When the first row is not available, returns an error.
+    pub fn first_row<'s, R: DeserializeRow<'s>>(&'s self) -> Result<R, FirstRowError> {
+        match self.maybe_first_row::<R>() {
+            Ok(Some(row)) => Ok(row),
+            Ok(None) => Err(FirstRowError::RowsEmpty),
+            Err(RowsError::NotRowsResponse) => Err(FirstRowError::NotRowsResponse),
+            Err(RowsError::TypeCheckFailed(err)) => Err(FirstRowError::TypeCheckFailed(err)),
+        }
+    }
+
+    /// Returns `Option<R>` containing the first of a result.\
+    /// Fails when the query isn't of a type that could return rows, same as [`rows()`](QueryResult::rows).
+    pub fn maybe_first_row<'s, R: DeserializeRow<'s>>(&'s self) -> Result<Option<R>, RowsError> {
+        Ok(self.rows::<R>()?.next().transpose()?)
+    }
```

Shouldn't `FirstRowError` be returned from `maybe_first_row` function?

---

`FirstRowError` - no, it is too broad, it can represent some failures which cannot happen with `maybe_first_row`.

Maybe you meant to say that `maybe_first_row` should return an error if the response was not of Rows type, like the comment suggests? In that case I agree and will fix that.

---

I will add a `MaybeFirstRowError` struct.